File: src/org/jetbrains/plugins/scala/projectView/TypeDefinitionNode.java
package org.jetbrains.plugins.scala.projectView;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.projectView.ViewSettings;
import com.intellij.ide.projectView.impl.nodes.AbstractPsiBasedNode;
import com.intellij.ide.projectView.impl.nodes.ClassTreeNode;
import com.intellij.ide.util.treeView.AbstractTreeNode;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiElement;
import org.jetbrains.plugins.scala.lang.psi.api.statements.ScValue;
import org.jetbrains.plugins.scala.lang.psi.api.statements.ScVariable;
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.ScNamedElement;
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.ScTypedDefinition;
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.ScMember;
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.ScObject;
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.ScTemplateDefinition;
import org.jetbrains.plugins.scala.lang.psi.light.PsiClassWrapper;
import scala.collection.Iterator;
import scala.collection.Seq;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;

// TODO Convert to Scala & Refactor
class TypeDefinitionNode extends ClassTreeNode {
  public TypeDefinitionNode(ClassTreeNode delegate) {
    super(delegate.getProject(), delegate.getValue(), delegate.getSettings());
    if (delegate.getValue() instanceof ScTemplateDefinition) {
      myName = ((ScTemplateDefinition) delegate.getValue()).name();
    }
  }

  @Override
  public String getTitle() {
    PsiClass value = getValue();
    if (value != null && value.isValid()) {
      if (value instanceof ScTemplateDefinition) {
        return ((ScTemplateDefinition) value).qualifiedName();
      } else return value.getQualifiedName();
    } else return null;
  }

  @Override
  public void updateImpl(PresentationData data) {
    final PsiClass aClass = getValue();
    if (aClass != null && aClass instanceof ScTemplateDefinition) {
      data.setPresentableText(((ScTemplateDefinition) aClass).name());
    } else super.updateImpl(data);
  }

  @Override
  public Collection<AbstractTreeNode> getChildrenImpl() {
    if (!getSettings().isShowMembers()) return super.getChildrenImpl();
    ArrayList<AbstractTreeNode> result = new ArrayList<AbstractTreeNode>();
    PsiClass value = getValue();
    if (value != null && value.isValid()) {
      if (value instanceof ScTemplateDefinition) {
        ScTemplateDefinition definition = (ScTemplateDefinition) value;
        Seq<ScMember> members = definition.members();
        Iterator<ScMember> iterator = members.iterator();
        while (iterator.hasNext()) {
          ScMember member = iterator.next();
          if (member instanceof PsiClass) {
            result.add(new TypeDefinitionNode(new ClassTreeNode(getProject(), (PsiClass) member, getSettings())));
          } else if (member instanceof ScNamedElement) {
            result.add(new ScNamedElementTreeNode(getProject(), (ScNamedElement) member, getSettings()));
          } else if (member instanceof ScValue) {
            ScValue v = (ScValue) member;
            Iterator<ScTypedDefinition> declared = v.declaredElements().iterator();
            while (declared.hasNext()) {
              ScTypedDefinition typed = declared.next();
              result.add(new ScNamedElementTreeNode(getProject(), typed, getSettings()));
            }
          } else if (member instanceof ScVariable) {
            ScVariable v = (ScVariable) member;
            Iterator<ScTypedDefinition> declared = v.declaredElements().iterator();
            while (declared.hasNext()) {
              ScTypedDefinition typed = declared.next();
              result.add(new ScNamedElementTreeNode(getProject(), typed, getSettings()));
            }
          }
        }
      }
    }
    return result;
  }

  private class ScNamedElementTreeNode extends AbstractPsiBasedNode<ScNamedElement> {
    public ScNamedElementTreeNode(Project project, ScNamedElement function, ViewSettings settings) {
      super(project, function, settings);
    }

    @Override
    protected PsiElement extractPsiFromValue() {
      return getValue();
    }

    @Override
    protected Collection<AbstractTreeNode> getChildrenImpl() {
      return Collections.emptyList();
    }

    @Override
    protected void updateImpl(PresentationData data) {
      ScNamedElement namedElement = getValue();
      if (namedElement != null) {
        String text = namedElement.name();
        data.setPresentableText(text);
      }
    }
  }

  @Override
  public PsiClass getPsiClass() {
    PsiClass psiClass = super.getPsiClass();
    if (psiClass instanceof ScObject) {
      ScObject obj = (ScObject) psiClass;
      return new PsiClassWrapper(obj, obj.qualifiedName(), obj.name());
    } else {
      return psiClass;
    }
  }
}

File: src/org/jetbrains/plugins/scala/projectView/TypeDefinitionNode.scala
package org.jetbrains.plugins.scala.projectView

import java.util

import com.intellij.ide.projectView.impl.nodes.ClassTreeNode
import com.intellij.ide.projectView.{PresentationData, ViewSettings}
import com.intellij.psi.PsiClass
import org.jetbrains.plugins.scala.lang.psi.api.statements.{ScValue, ScVariable}
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.ScNamedElement
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.ScTypeDefinition
import org.jetbrains.plugins.scala.lang.psi.light.PsiClassWrapper
import org.jetbrains.plugins.scala.project.ProjectContext

import scala.collection.JavaConverters._

/**
  * @author Pavel Fatin
  */
private class TypeDefinitionNode(definition: ScTypeDefinition)(implicit project: ProjectContext, settings: ViewSettings)
  extends ClassTreeNode(project, definition, settings) {

  myName = definition.name

  override def getTitle: String =
    value.map(_.qualifiedName).getOrElse(super.getTitle)

  override def updateImpl(data: PresentationData): Unit = value match {
    case Some(it) => data.setPresentableText(it.name)
    case None => super.updateImpl(data)
  }

  override def getChildrenImpl: util.Collection[Node] =
    if (getSettings.isShowMembers) value.map(childrenOf).getOrElse(Seq.empty).asJava
    else super.getChildrenImpl

  private def childrenOf(parent: ScTypeDefinition): Seq[Node] = parent.members.flatMap {
    case definition: ScTypeDefinition =>
      Seq(new TypeDefinitionNode(definition))

    case element: ScNamedElement =>
      Seq(new NamedElementNode(element))

    case value: ScValue =>
      value.declaredElements.map(new NamedElementNode(_))

    case variable: ScVariable =>
      variable.declaredElements.map(new NamedElementNode(_))

    case _ => Seq.empty
  }

  override def getPsiClass: PsiClass =
    value.filter(_.isObject)
      .map(it => new PsiClassWrapper(it, it.qualifiedName, it.name))
      .getOrElse(super.getPsiClass)

  private def value: Option[ScTypeDefinition] = Some(getValue) collect {
    case value: ScTypeDefinition if value.isValid => value
  }
}
